src/main/java/frc/robot/poseestimator/PoseEstimatorMath.java

```diff
@@ -0,0 +1,94 @@
+package frc.robot.poseestimator;
+
+import edu.wpi.first.math.MathUtil;
+import edu.wpi.first.math.geometry.Pose2d;
+import edu.wpi.first.math.geometry.Rotation2d;
+import edu.wpi.first.math.geometry.Transform2d;
+import edu.wpi.first.math.geometry.Twist2d;
+import frc.robot.poseestimator.observations.VisionObservation;
+
+public class PoseEstimatorMath {
+
+	public static Twist2d addGyroToTwist(Twist2d twist, Rotation2d currentGyroAngle, Rotation2d lastGyroAngle) {
+		boolean hasGyroUpdated = currentGyroAngle != null;
+		if (hasGyroUpdated) {
+			return updateChangeInAngle(twist, currentGyroAngle, lastGyroAngle);
+		}
+		return twist;
+	}
+
+	public static Twist2d updateChangeInAngle(Twist2d twist, Rotation2d currentGyroAngle, Rotation2d lastGyroAngle) {
+		double rotationDifference = currentGyroAngle.getRadians() - lastGyroAngle.getRadians();
+		double wrappedRotationDifference = MathUtil.angleModulus(rotationDifference);
+		return new Twist2d(twist.dx, twist.dy, wrappedRotationDifference);
+	}
+
+	public static double[] combineStandardDeviations(double[] odometryStandardDeviations, double[] visionStandardDeviations) {
+		double[] combinedStandardDeviations = new double[3];
+		for (int i = 0; i < combinedStandardDeviations.length; i++) {
+			double odometryStandardDeviation = odometryStandardDeviations[i];
+			double visionStandardDeviation = visionStandardDeviations[i];
+			combinedStandardDeviations[i] = combineStandardDeviations(odometryStandardDeviation, visionStandardDeviation);
+		}
+		return combinedStandardDeviations;
+	}
+
+	public static double combineStandardDeviations(double odometryStandardDeviation, double visionStandardDeviation) {
+		if (odometryStandardDeviation == 0) {
+			return 0;
+		}
+		double squaredVisionStandardDeviation = visionStandardDeviation * visionStandardDeviation;
+		return odometryStandardDeviation
+			/ (odometryStandardDeviation + Math.sqrt(odometryStandardDeviation * squaredVisionStandardDeviation));
+	}
+
+	public static double[] multiplyArrays(double[] array1, double[] array2) {
+		double[] result = new double[Math.min(array1.length, array2.length)];
+		for (int i = 0; i < result.length; i++) {
+			result[i] = array1[i] * array2[i];
+		}
+		return result;
+	}
```

Comment:
It'll probably be more efficient to preform it as a matmul of a horizontal and a vertical matrices. idk though

Reply:
it will not